### DISCUSSION TITLE
Issues with setting up Pi Zero

### DISCUSSION THREAD

**Discussion by eggyclaw** [Berny23/LD-ToyPad-Emulator#47] [Q&A]
I've been having a number of problems setting up a Pi Zero W2.

If I follow the instructions under option 2:
1. fresh install of Raspberry Pi OS lite on a 2gb SD card.
2.  Opened the 2 links provided to set up headless mode (I did not do all steps from these guides and only focused on the sections that mentioned 'headless' mode.  Please clarify if this was correct)
- add an 'ssh' file to the boot drive of the SD card to activate ssh
- add an 'wpa_supplicant.conf ' file to the bood drive of the SD card to set up wifi
3. put SD card in Pi and connect to PC using the 'USB' socket
4.  Copy and paste the scripts in option 2 in the order given
5. launch the emulator
6. emulator runs on webpage but cannot detect game in CEMU
7. PC (Win 10) does not seem to register a USB device being attached.   ZADig does not show anything.

I've also tried a few other things that wasn't part of your instructions but unsure whether these are needed - either way it hasn't helped!

- use the following guide to turn of USB Gadget mode https://learn.adafruit.com/turning-your-raspberry-pi-zero-into-a-usb-gadget/ethernet-gadget
- Doing this allowed my PC and ZADig to detect a device - albeit as a Ethernet gadget.
- NPM was giving warnings that utf-8-validate and bufferutil were not installed.   'npm install' dumped lots of things in the log on first run.   After I installed these 2 npm plugins 'npm install' output was only 2 lines.

These actually ended up breaking node index.js, so I assume these additions aren't required. If of interest the error from index.js is below.  
 
Are you able to help clarify where I might be going wrong?

> fs.js:119
>     throw err;
>     ^
> 
> Error: ENOENT: no such file or directory, open '/dev/hidg0'
>     at Object.openSync (fs.js:448:3)
>     at new RawTransport (/home/pi/LD-ToyPad-Emulator/node_modules/node-ld/dist/transports/raw.js:11:27)
>     at new ToyPadEmu (/home/pi/LD-ToyPad-Emulator/node_modules/node-ld/dist/lib/ToyPadEmu.js:65:69)
>     at Object.<anonymous> (/home/pi/LD-ToyPad-Emulator/index.js:24:10)
>     at Module._compile (internal/modules/cjs/loader.js:816:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:827:10)
>     at Module.load (internal/modules/cjs/loader.js:685:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:620:12)
>     at Function.Module.runMain (internal/modules/cjs/loader.js:877:12)
>     at internal/main/run_main_module.js:21:11

**Comment by Berny23** [comment]
Never use Zadig with Cemu, it breaks everything.

This error means that you didn't correctly install the usb gadget via the script included on the guide. Make sure to copy and paste every single command.

Don't follow other guides about usb gadgets, it was already complicated enough to write this guide to get it working in the first place. Making any additional changes to these usb files will break their functionality.

Changes to the emulator's files are not necessary, if you don't know what they do.

I suggest you rollback the driver, remove Zadig and start over with the guide.
Alternatively, try another usb port on the Raspberry Pi and another usb cable (that supports data transfer, like for phones!).

**Reply by eggyclaw** [reply]
Thanks for responding so quickly Berny23.

To clarify on my previous post - I initially followed the guide step by step twice.   Both times with clean fresh SD images to ensure I wasn't overwriting anything.   First time I pasted the script in SSH as blocks of code then executed, the second time I pasted and executed line by line checking for any errors etc.   Both times the web server worked but my PC and CEMU did not pick up the USB input.

The only question I still have on following the guide is to ask for clarification on what instructions from the 2 guides you mention should be followed?

> If you're using a Raspberry Pi Zero W, flash Raspberry Pi OS Lite to your SD card using the [Raspberry Pi Imager tool](https://www.raspberrypi.org/software/) and follow [this ](https://www.raspberrypi.org/documentation/configuration/wireless/headless.md)as well as [this ](https://www.raspberrypi.org/documentation/remote-access/ssh/README.md)instruction for headless installation.

Also to confirm - I've not used ZaDig to install a driver - it was only used to see if it could find and list the USB gadget, which it did not find.

Some extra info in case it's useful:
- From Raspberry Pi Imager - used Raspberry Pi OS Lite (32 bit) Released 2021-10-30
- SSH using Windows 10 cmd
- SSH as Pi@ IP Address
- Logged in using Pi default password
- I was not prompted for any other password/authentication when running the script

**Reply by Berny23** [reply]
Yes, you only have to follow the parts which mention headless installation, like I wrote: "instructions for headless installation"

What is the output of `lsb_release -d`?
What is the output of `tail /usr/local/bin/toypad_usb_setup.sh`?
What is the output of `sudo crontab -l`?

Did you try other USB cables? Don't use an OTG cable!
First, test your USB cable by connecting your phone to your PC with it; if you're able to transfer files, it will also work with the emulator.

You must use the port labelled "USB" on this picture:
![image](https://user-images.githubusercontent.com/36038743/151584326-aafaa34c-bf5e-4bc8-ad71-7355aac0808c.png)

**Reply by eggyclaw** [reply]
Thanks for a quick response again.
After some playing about it turns out there was something off with the 2Gb SD card I was using.   All of a sudden on a fresh wipe it started saying there was no space left to complete the install etc etc.

I dug out a 16Gb card and ran through the guide and all was well.

So note for all - if something isn't quite right it could be either a bad SD card or you need more than 2Gb.
